// blinktrade/lib/constants/actionTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var BALANCE = exports.BALANCE = 'BALANCE';

var ORDER_BOOK = exports.ORDER_BOOK = 'OB';
var ORDER_BOOK_TRADE_NEW = exports.ORDER_BOOK_TRADE_NEW = 'TRADE_NEW';
var ORDER_BOOK_NEW_ORDER = exports.ORDER_BOOK_NEW_ORDER = 'NEW_ORDER';
var ORDER_BOOK_UPDATE_ORDER = exports.ORDER_BOOK_UPDATE_ORDER = 'UPDATE_ORDER';
var ORDER_BOOK_DELETE_ORDER = exports.ORDER_BOOK_DELETE_ORDER = 'DELETE_ORDER';
var ORDER_BOOK_DELETE_ORDERS_THRU = exports.ORDER_BOOK_DELETE_ORDERS_THRU = 'DELETE_ORDERS_THRU';

var EXECUTION_REPORT = exports.EXECUTION_REPORT = 'EXECUTION_REPORT';
var EXECUTION_REPORT_NEW = exports.EXECUTION_REPORT_NEW = 'NEW';
var EXECUTION_REPORT_PARTIAL = exports.EXECUTION_REPORT_PARTIAL = 'PARTIAL';
var EXECUTION_REPORT_EXECUTION = exports.EXECUTION_REPORT_EXECUTION = 'EXECUTION';
var EXECUTION_REPORT_CANCELED = exports.EXECUTION_REPORT_CANCELED = 'CANCELED';
var EXECUTION_REPORT_REJECTED = exports.EXECUTION_REPORT_REJECTED = 'REJECTED';

var DEPOSIT_REFRESH = exports.DEPOSIT_REFRESH = 'DEPOSIT_REFRESH';
var WITHDRAW_REFRESH = exports.WITHDRAW_REFRESH = 'WITHDRAW_REFRESH';

/* eslint-disable quote-props */
var EVENTS = exports.EVENTS = {
  ORDERBOOK: {
    '0': ORDER_BOOK_NEW_ORDER,
    '1': ORDER_BOOK_UPDATE_ORDER,
    '2': ORDER_BOOK_DELETE_ORDER,
    '3': ORDER_BOOK_DELETE_ORDERS_THRU
  },
  TRADES: {
    '0': ORDER_BOOK_TRADE_NEW
  },
  EXECUTION_REPORT: {
    '0': EXECUTION_REPORT_NEW,
    '1': EXECUTION_REPORT_PARTIAL,
    '2': EXECUTION_REPORT_EXECUTION,
    '4': EXECUTION_REPORT_CANCELED,
    '8': EXECUTION_REPORT_REJECTED
  }
};
